wasm/_utils/singleton.py

```diff
@@ -0,0 +1,16 @@
+from typing import (
+    Any,
+    Dict,
+    Type,
+)
+
+
+class Singleton:
+    _cache: Dict[Type['Singleton'], 'Singleton'] = {}
+
+    def __new__(cls, *args: Any, **kwargs: Any) -> 'Singleton':
+        if cls not in cls._cache:
+            instance = super().__new__(cls)
+            instance.__init__(*args, **kwargs)
```

Comment:
I think there's a foot-gun here in that the result of `__new__()` [will have `__init__()` called upon it](https://docs.python.org/3/reference/datamodel.html#object.__new__). For the first instance of the singleton `__init__()` is called twice! Subsequent calls to the Singleton will return the same object, but then `__init__()` will be called on that object. Code which assumes that "single instance" means "single initialization" will be buggy, and I think that's a reasonable assumption to make.

Reply:
Fixed in https://github.com/ethereum/py-wasm/pull/38